extensions/amp-story/0.1/parallax.js

```diff
@@ -0,0 +1,256 @@
+          let zIndex = 1;
+          const layers = this.getLayers(page.element);
+          layers.map(layer => {
+            const fxElement = new ParallaxElement(
+                dev().assertElement(layer),
+                zIndex++,
+                layers.length
+            );
+            this.parallaxElements_.push(fxElement);
+          });
+        });
+
+    listen(global, 'deviceorientation', event => {
+      this.parallaxOrientationMutate_(
+          event,
+          this.parallaxElements_,
+          viewport
+      );
+    });
+  }
+
+  /**
+   * Discovers and returns all layers inside a page
+   * @param {Element} page
+   * @return {Array<Element>}
+   */
+  getLayers(page) {
+    return toArray(page.querySelectorAll('amp-story-grid-layer'));
+  }
+
+  /**
+   * Calculate the smoothed change in position using the device orientation
+   * change event then update each of the layers will its calculated position.
+   * @param {Event} event
+   * @param {!Array<!ParallaxElement>} elements
+   * @param {!../../../src/service/viewport/viewport-impl.Viewport} viewport
+   * @private
+   */
+  parallaxOrientationMutate_(event, elements, viewport) {
+    const window = this.win_;
+    const screen = window.screen;
+    let gamma = event.gamma;
+    let beta = event.beta;
+    let angle;
+
+    // Detect the implementation of orientation angle
+    if ('orientation' in screen)
+    {
+      angle = screen.orientation.angle;
+    } else {
+      angle = window.orientation;
+    }
+
+    // Reverse gamma/beta if the device is in landscape
+    if (window.orientation == 90 || window.orientation == -90) {
+      const tmp = gamma;
+      gamma = beta;
+      beta = tmp;
+    }
+
+    // Flip signs of the angles if the phone is in 'reverse landscape' or
+    // 'reverse portrait'
+    if (angle < 0) {
+      gamma = -gamma;
+      beta = -beta;
+    }
+
+    // Smooth the gamma value (X-AXIS)
+    if (this.smoothingPointsX_.length > SMOOTHING_PTS) {
+      this.smoothingPointsX_.shift();
+    }
+    this.smoothingPointsX_.push(gamma);
+    const avgX = sum(this.smoothingPointsX_) / SMOOTHING_PTS;
+    if (this.smoothingPointsX_.length > SMOOTHING_PTS && this.middleX_ == 0) {
+      this.middleX_ = avgX;
+    }
+
+    // Smooth the beta value (Y-AXIS)
+    if (this.smoothingPointsY_.length > SMOOTHING_PTS) {
+      this.smoothingPointsY_.shift();
+    }
+    this.smoothingPointsY_.push(beta);
+    const avgY = sum(this.smoothingPointsY_) / SMOOTHING_PTS;
+    if (this.smoothingPointsY_.length > SMOOTHING_PTS && this.middleY_ == 0) {
+      this.middleY_ = avgY;
+    }
+
+    let mappedX = avgX - this.middleX_;
+    let mappedY = avgY - this.middleY_;
+
+    // Limit the range for a smoother/less shaky effect
+    mappedX = mapRange(mappedX, -75, 75, -25, 25);
+    mappedY = mapRange(mappedY, -75, 75, -25, 25);
+
+    elements.forEach(element => {
+      if (element.shouldUpdate(viewport)) {
```

Comment:
`shouldUpdate` should be in a vsync measure
`update` should be in a vsyn mutate